repo2docker/contentproviders/hydroshare.py

```diff
@@ -0,0 +1,110 @@
+import zipfile
+import os
+import shutil
+import time
+import json
+from datetime import datetime, timezone, timedelta
+
+from urllib.request import urlopen, Request, urlretrieve
+from urllib.error import HTTPError
+
+from .doi import DoiProvider
+from .base import ContentProviderException
+from ..utils import normalize_doi, is_doi
+
+
+class Hydroshare(DoiProvider):
+    """Provide contents of a Hydroshare resource."""
+
+    def detect(self, doi, ref=None, extra_args=None):
+        """Trigger this provider for things that resolve to a Hydroshare resource"""
+        hosts = [
+            {
+                "hostname": [
+                    "https://www.hydroshare.org/resource/",
+                    "http://www.hydroshare.org/resource/",
+                ],
+                "django_irods": "https://www.hydroshare.org/django_irods/download/bags/",
+                "version": "https://www.hydroshare.org/hsapi/resource/{}/scimeta/elements",
+            }
+        ]
+
+        def fetch_version(resource_id, host):
+            """Fetch resource modified date and convert to epoch"""
+            json_response = json.loads(
+                self.urlopen(host["version"].format(self.resource_id)).read()
+            )
+            date = next(
+                item for item in json_response["dates"] if item["type"] == "modified"
+            )["start_date"]
+            # Hydroshare timestamp always returns the same timezone, so strip it
+            date = date.split(".")[0]
+            parsed_date = datetime.strptime(date, "%Y-%m-%dT%H:%M:%S")
+            epoch = parsed_date.replace(tzinfo=timezone(timedelta(0))).timestamp()
+            # truncate the timestamp
+            return str(int(epoch))
+
+        url = self.doi2url(doi)
```

Comment:
The indentation of this and the following code block is wild. Should it be indented by one more level? If it is here on purpose then we need to move it as class level code between two methods is not something we do ;)

Reply:
Ah, it turns out this is actually the right level of indentation but I got distracted by `def fetch_version`'s indentation. is there a reason to have that function inline like that? We could just make it a method. But this is nitpicking